### PR TITLE
feat: Factory Deployment Script

### PR DESCRIPTION
**Pull request by Zer0dot**
## Motivation

We need scripts to predict the factory address and to deploy it. 

## Solution

Implement the aforementioned scripts, as well as some minor refactors to reduce resulting code duplication.

## TODO

- [x] Create the factory address prediction script.
- [x] Test the deployment

### CODE REVIEW

**Comment by Zer0dot** [comment]
Note that the factory address prediction reads addresses from the env for dependencies, rather than using the predicted addresses in the same script, as those may differ from the predicted addresses.

File: script/DeployFactory.s.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        entryPoint = _getEntryPoint();
+        modularAccountImpl = _getModularAccountImpl();
+        semiModularAccountBytecodeImpl = _getSemiModularAccountBytecodeImpl();
+        singleSignerValidationModule = _getSingleSignerValidationModule();
+        webAuthnValidationModule = _getWebAuthnValidationModule();
+        factoryOwner = _getFactoryOwner();
+
+        // Load the expected address and salt from env vars.
+        expectedFactoryAddr = vm.envOr("FACTORY", address(0));
+        factorySalt = vm.envOr("FACTORY_SALT", uint256(0));
```
Comment (adamegyed) [inline script/DeployFactory.s.sol]: I think the current var used in the `.env.example` file is `ACCOUNT_FACTORY` and `ACCOUNT_FACTORY_SALT`. I'm fine with using either that or the one here, but could you make it consistent across the locations?

File: script/GetInitcodeHash.s.sol [hunk 2]
```diff
@@ -47,16 +47,8 @@ contract GetInitcodeHashScript is ScriptBase, Artifacts {
...
-        if (address(entryPoint) == address(0)) {
-            console.log(
-                "Env Variable 'ENTRYPOINT' not found or invalid, defaulting to v0.7 EntryPoint at "
-                "0x0000000071727De22E5E9d8BAf0edAc6f37da032"
-            );
-            entryPoint = IEntryPoint(0x0000000071727De22E5E9d8BAf0edAc6f37da032);
-        } else {
-            console.log("Using user-defined EntryPoint at: %x", address(entryPoint));
-        }
+        IEntryPoint entryPoint = _getEntryPoint(); //IEntryPoint(payable(vm.envOr("ENTRYPOINT", address(0))));
```
Comment (adamegyed) [inline script/GetInitcodeHash.s.sol]: nit: delete comment

File: script/ScriptBase.sol [hunk 3]
```diff
@@ -17,6 +20,41 @@ abstract contract ScriptBase is Script {
...
+
+        return entryPoint;
+    }
+
+    function _getModularAccountImpl() internal view returns (ModularAccount) {
+        return ModularAccount(payable(vm.envOr("MODULAR_ACCOUNT_IMPL", address(0))));
+    }
+
+    function _getSemiModularAccountBytecodeImpl() internal view returns (SemiModularAccountBytecode) {
+        return SemiModularAccountBytecode(payable(vm.envOr("SEMI_MODULAR_ACCOUNT_BYTECODE_IMPL", address(0))));
```
Comment (adamegyed) [inline script/ScriptBase.sol:43]: I think there's a different format already set up in `.env.example` without the `_IMPL` ending, could you make that consistent here too?

File: test/script/DeployFactory.s.t.sol [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    function setUp() public {
+        _deployFactoryScript = new DeployFactoryScript();
+
+        bytes32 zeroSalt = bytes32(0);
+
+        entryPoint = IEntryPoint(address(6));
+        modularAccountImpl = ModularAccount(payable(address(1)));
+        semiModularAccountBytecodeImpl = SemiModularAccountBytecode(payable(address(2)));
+        singleSignerValidationModule = address(3);
+        webAuthnValidationModule = address(4);
```
Comment (adamegyed) [inline test/script/DeployFactory.s.t.sol]: Why not deploy real copies of them? The EP can use the deploy script in OptimizedTest.

File: test/script/DeployFactory.s.t.sol [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        factoryOwner = address(5);
```
Comment (adamegyed) [inline test/script/DeployFactory.s.t.sol]: nit: prefer `makeAddr` over incrementing numbers due to appearing as precompiles in the test traces.

**Review by adamegyed — APPROVED** [review]
Looks good, had a couple small comments

File: test/script/DeployFactory.s.t.sol (hunk 4 above)
Comment (Zer0dot) [reply test/script/DeployFactory.s.t.sol]: Honestly, could go either way, I just didn't really want to test anything related to deploying anything else-- but it's a small change, I'll address this!

File: script/ScriptBase.sol (hunk 3 above)
Comment (Zer0dot) [reply script/ScriptBase.sol:43]: Yup, will do!

File: script/DeployFactory.s.sol (hunk 1 above)
Comment (Zer0dot) [reply script/DeployFactory.s.sol]: Will do! Good catch, in my infinite wisdom I didn't check .env.example :)

File: script/ScriptBase.sol (hunk 3 above)
Comment (Zer0dot) [reply script/ScriptBase.sol:43]: Oops-- actually the one in `.env.example` has `_IMPL`, I'd say we're good to keep it, it ensures the developer knows we're looking for an implementation, not an actual account.

**Pull request merged**